cqrs-postgres/src/store.rs

```diff
@@ -287,6 +288,107 @@ where
         ])?;
         Ok(rows.iter().map(|r| r.get(0)).collect())
     }
+
+    /// Reads events and associated metadata from the event source for a given identifier.
+    ///
+    /// Only loads events after the event number provided in `since` (See [Since]), and will only load a maximum of
+    /// `max_count` events, if given. If not given, will read all remaining events.
+    pub fn read_events_with_metadata<I>(
+        &self,
+        id: &I,
+        since: Since,
+        max_count: Option<u64>,
+    ) -> Result<
+        Option<Vec<Result<VersionedEventWithMetadata<E, M>, LoadError<E::Error>>>>,
+        LoadError<E::Error>,
+    >
+    where
+        I: AggregateId<A>,
+        E: DeserializableEvent,
+        M: for<'de> serde::Deserialize<'de>,
+    {
+        let last_sequence = match since {
+            cqrs_core::Since::BeginningOfStream => 0,
+            cqrs_core::Since::Event(x) => x.get(),
+        } as i64;
+
+        let events;
+        let trans = self
+            .conn
+            .transaction_with(postgres::transaction::Config::default().read_only(true))?;
+
+        let handle_row = |row: postgres::rows::Row| {
+            let sequence: Sequence = row.get(0);
+            let event_type: String = row.get(1);
+            let raw: RawJsonRead = row.get(2);
+            let metadata: Json<M> = row.get(3);
+            let event = E::deserialize_event_from_buffer(&raw.0, &event_type)
+                .map_err(LoadError::DeserializationError)?
+                .ok_or_else(|| LoadError::UnknownEventType(event_type.clone()))?;
+            log::trace!(
+                "entity {}: loaded event; sequence: {}, type: {}",
+                id.as_str(),
+                sequence.0,
+                event_type
+            );
+            Ok(VersionedEventWithMetadata {
+                sequence: sequence.0,
+                event,
+                metadata: metadata.0,
+            })
+        };
+
+        let stmt;
+        {
+            let mut rows;
+            if let Some(max_count) = max_count {
+                stmt = trans.prepare_cached(
+                    "SELECT sequence, event_type, payload, metadata \
+                     FROM events \
+                     WHERE aggregate_type = $1 AND entity_id = $2 AND sequence > $3 \
+                     ORDER BY sequence ASC \
+                     LIMIT $4",
+                )?;
+                rows = stmt.lazy_query(
+                    &trans,
+                    &[
+                        &A::aggregate_type(),
+                        &id.as_str(),
+                        &last_sequence,
+                        &(max_count.min(i64::max_value() as u64) as i64),
```

Comment:
This line is pretty dense with conversions, anyway this could be made more legible?

perhaps pull part out to a line above
`let default_max_coult = i64::max_value() as u64;`

Reply:
I found a nicer way to do this [here][1] in #4. I'll probably make a few changes here to make it look more like that PR.

  [1]: https://github.com/cq-rs/cqrs/pull/4/files#diff-d865866cae8062948f6fc7ac17cf00f7R423